daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/RecordWalker.scala

```diff
@@ -0,0 +1,309 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dsom.walker
+
+import java.io.File
+import java.util
+
+import org.apache.daffodil.dsom.{Choice, ComplexTypeBase, ElementBase, GroupRef, Root, Sequence, SimpleTypeBase, Term}
+import org.apache.nifi.serialization.SimpleRecordSchema
+import org.apache.nifi.serialization.record.{DataType, RecordField, RecordFieldType, RecordSchema, SchemaIdentifier}
+import org.apache.nifi.serialization.record.`type`.RecordDataType
+import scala.collection.mutable.ListBuffer
+
+/**
+ * Direct subclass of the NIFI RecordField.  NIFI doesn't have "Optional" fields, so
+ * eventually this will either be removed from the Schema if an Infoset doesn't have the field,
+ * or it will become a regular RecordField if it does.
+ * @param fieldName the name for the RecordField
+ * @param dataType the NIFI DataType of this RecordField
+ */
+class OptionalRecordField(fieldName: String, dataType: DataType)
+  extends RecordField(fieldName, dataType) {
+  // this is the easiest way to construct it; essentially a "copy constructor" except to denote
+  // that we want to treat it as this special optional subclass.
+  def this(recordField: RecordField) = this(recordField.getFieldName, recordField.getDataType)
+  override def equals(obj: Any): Boolean =
+    obj match {
+      case _: OptionalRecordField => super.equals(obj)
+      case _ => false
+    }
+  override def hashCode(): Int = 31 * super.hashCode() + 1
+  override def toString: String = "Optional" + super.toString
+}
+
+/**
+ * Concrete implementation of the AbstractDSOMWalker abstract class.
+ * This class produces a NIFI RecordSchema that is intended to match the original DFDL file.
+ *
+ * The RecordSchema is built in 3 primary stages:
+ * 1) A tree of SchemaNodes is created as the DFDL file is walked; this walk is performed
+ * through the various event handlers defined in the parent abstract class.
+ * 2) The tree of SchemaNodes undergoes some post-processing, mainly to remove redundant Record wrappers.
+ * 3) The tree of SchemaNodes is converted into a RecordSchema; it is walked recursively within this class.
+ * @param schemaFile an input DFDL Schema file to parse
+ */
+class RecordWalker(schemaFile: File) extends AbstractDSOMWalker[RecordSchema](schemaFile) {
```

Comment:
This is pretty great! I'm surprised at how little code is needed to hook this all together.

However, one thing we need to discuss is where this code actually lives. I think it's safe to safe it shouldn't live in the daffodil-core module. If it does, that means anyone depending on daffodil-core will now pull in a NiFi dependency, which shouldn't be necessary. The DSOM related stuff makes sense to be in core, because it needs to know how to walk the DSOM. But the NiFi specifc code needs to be in a separate module for dependency purposes.

The question, is where does that separate module live? One option would be to create an "integrations" directory, and that could contain separate directories for different kinds of integrations. The jars created from those modules can have separate dependencies so you only need to pull in those dependencies if you're using the integration jars. Something like this probably makes the most sense.

Another thing that plays a role where this lives is this RecordWalker is just one part of NiFi integration. There also needs to be a NiFi Controller/Processor that use this RecordWalker. Ideally, that too would live in the same module so that the result of this module could be a NiFi nar. One issue with that is generally maven is used to build special NiFi nars. Integrating that into the sbt build process might pose some challenges, especially since we have a couple sbt plugins. But maybe we don't want that integration so that building Daffodil doesn't require pulling in dependencies from a bunch of different frameworks? We could of course include those in our github actions, but perhaps they aren't build by default?

Another option is to put integrations in separate repos. That makes integration tests a bit more complicated. And I'm not sure how much benefit there is. The clear separate is maybe nice, but perhaps that's too much separation.

Note that there also exists Daffodil NiFi processor: https://github.com/TresysTechnology/nifi-daffodil

We probably want to determine how that might integrate in this as well. Perhaps it can all go in a single nar?

Another option, maybe we should attempt to upstream this into NiFi and the only thing that exists in this repo are the abstract walker and the API changes (plus regression tests not specific to NiFi). There is already an open pull request to add the Nif Processor here: https://github.com/apache/nifi/pull/3130. The NiFi devs didn't seem adverse to Daffodil support, but wondering if Records is the better approch. So they may be open to accepting this kind of change.

Reply:
I agree, I definitely want to separate all of the NIFI-related code from anything related to DSOM.  I am working on NIFI Processors that use this entire DSOM API but they (more so the Unparse processor) are not quite finished yet, I am still doing a great deal of refactoring.  

I am not sure that any of the NIFI code either than RecordWalker and RecordSchemaNode need exist in the main Daffodil codebase - if it were to go here, would it still go in core but just a different subdirectory, or would the japi/sapi modules be more appropriate?  

Since the base of my code largely builds off of the repo you linked, let me know if you think it is appropriate if I simply rebase my project onto that one or if we should go with a different set of processors since they are a good bit different when they utilize NIFI Records internally.

Reply:
In theory, the only thing that needs to be in the Daffodil codebase is the API, which is essentially the AbstractDSOMWalker and all the Views traits. Nothing depending on NiFi should go in daffodil-core because that means core depends on NiFi and thus anyone using Daffodil now needs a NiFi jar--the use of Daffodli shouldn't depend on NiFi. 

I don't think it really matters where the AbstractDSOMWalker and View traits go. They can probably just stay in core. Note that they can't go in the japi/sapi directories. This is because daffodilxXapi currently depends on daffodil-core (xapi is essentially a basic wrapper around core). If these AbstractDSOMWalker and View traits were in the daffodil-xapi, then daffodil-core would then depend on daffodil-xapi, so we'd have a circular dependency. So the AbstractDSOMWalker and View traits either stay in daffodil-core or in a new subproject. Since we are considering this experiemental, maybe it makes sense to keep it in core, and when we finalize an API and move it from experimental, we can move it to a separate dir where we can generate public java/scaladocs? I guess this is essentially what we've done with daffodil-udf. I consider it a public API so it makes sense to get it's own directory where we can generate docs for it.

So where does everything go that uses NiFi? It should probably all stick together in one place. Perhaps that place right now is the nifi-daffodil repo, just to make it very obvious what is NiFi and what is Daffodil, and to highlight what changes are Daffodil API additions and what are NiFi usages of this API? Once that's finalized and the nifi-daffodil repo is in a stable state and the experimental API is considered sufficient, it then becomes pretty trivial to either merge that into Nifi if they are willing to accept it, or merge it into a new Daffodli "integrations" directory for it to live. Or it could just remain separate--though I'd prefer it get merged somewhere.

Reply:
It definitely makes sense to keep the API and any NiFi-related/dependent code separate; in the latest revision I removed RecordWalker and SchemaRecordNode from core and made some general classes for testing to replace them.  I think it would make the most sense to move both of these to the nifi-daffodil repo, but as it currently stands they're both written in Scala and I'm not sure if there's necessarily a 1-1 Java conversion.

Reply:
Now that the walk is initiated externally from `ProcessorFactory` rather than from the walker itself, there is currently no way to start it via the Java or Scala API.  I am going to go ahead and add the `walkDSOM` method to the `ProcessorFactory` in the JAPI and SAPI modules; let me know if you think something else should be done instead.

Reply:
Re: the nifi-daffodil repo, if you're able to modify the nifi-daffodil repo to support scala, it should work fine in theory. I suspect the new controllers are mostly separate from the existing processors?

Re: walkDSOM on ProcessorFactory, I think after discussions the consensus is that while a recursive walk is useful, there may be cases where it isn't sufficient. So I would suggust that we move all the ``walkDSOM`` functions off of the Views and onto the AbstractDSOMWalker (just like you had originally), so the AbstractDSOM walker has all the logic for how to recursively walk a DSOM. Then the only new function need on the the ProcessorFactory is an ability to get the RootView. If one wants to traverse this RootView manually, they can do so. But if they just want to traverse it recrusively and get events, they can create a concrete implementation of the AbstractDSOMWalker and implement those events (and potential override the walkDSOM function if they want. So for example:
```scala
val rootView = processorFactory.rootView
val walker = new DSOMWalkerImpl()
walker.walkDSOM(rootView)
```
So the View's don't contain any logic about how they should be traversed, since they don't really know that. Recurisve makes sense, but presumable there are reasonable traversals. And technically the AbstractDSOMWalker isn't even neccessary--the views have all the information needed to traverse them. But to be helpful, it makes sense to also provide a AbstractDSOMWalker interface that contains all the logic for traversing recursively so users don't have to think about that and can just think of it as a series of events.

So this is very similar to what you originally started with, but just with Views to hide internal details of our dsom objects.